src/QueryM.purs

```diff
@@ -652,29 +748,76 @@ mkOgmiosWebSocket' lvl serverCfg continue = do
-            evaluateTxPendingRequests
-        , getProtocolParameters: mkListenerSet
-            getProtocolParametersDispatchMap
-            getProtocolParametersPendingRequests
-        , submit: mkListenerSet submitDispatchMap submitPendingRequests
-        , eraSummaries:
-            mkListenerSet eraSummariesDispatchMap eraSummariesPendingRequests
-        , currentEpoch:
-            mkListenerSet currentEpochDispatchMap currentEpochPendingRequests
-        , systemStart:
-            mkListenerSet systemStartDispatchMap systemStartPendingRequests
-        }
+      continue (Right ogmiosWs)
   pure $ Canceler $ \err -> liftEffect do
     _wsClose ws
     continue $ Left $ err
   where
   logger :: LogLevel -> String -> Effect Unit
   logger = logString lvl
 
+-- | For all pending `SubmitTx` requests checks if a transaction was added
+-- | to the mempool or included in the block before retrying the request.
+resendPendingSubmitRequests
+  :: OgmiosWebSocket
+  -> DatumCacheWebSocket
+  -> LogLevel
+  -> (forall (inp :: Type). RequestBody /\ inp -> Effect Unit)
+  -> DispatchIdMap Ogmios.SubmitTxR
+  -> PendingRequests (TxHash /\ CborBytes)
+  -> Effect Unit
+resendPendingSubmitRequests ogmiosWs datumCacheWs lvl sendRequest dim pr = do
+  submitPendingRequests <- Ref.read pr
+  unless (Map.isEmpty submitPendingRequests) do
+    -- Acquiring a mempool snapshot should never fail and,
+    -- after ws reconnection, should be instantaneous.
+    withMempoolSnapshot ogmiosWs lvl case _ of
+      Nothing ->
+        liftEffect $ traverse_ sendRequest submitPendingRequests
+      Just ms -> do
+        -- A delay of 5 sec for transactions to be processed by the node
+        -- and added to the mempool:
+        delay (wrap 5000.0)
+        let (pr' :: Array _) = Map.toUnfoldable submitPendingRequests
+        for_ pr' \(listenerId /\ requestBody /\ requestInput) ->
+          case requestInput of
+            Nothing ->
+              liftEffect $ sendRequest (requestBody /\ unit)
+            Just (txHash /\ _) -> do
+              handlePendingSubmitRequest ms listenerId requestBody txHash
+  where
+  handlePendingSubmitRequest
+    :: Ogmios.MempoolSnapshotAcquired
+    -> ListenerId
+    -> RequestBody
+    -> TxHash
+    -> Aff Unit
+  handlePendingSubmitRequest ms listenerId requestBody txHash = do
+    -- Check if the transaction was added to the mempool:
+    txInMempool <- mempoolSnapshotHasTxAff ogmiosWs lvl ms txHash
+    logger "Tx in the mempool" txInMempool txHash
+    retrySubmitTxRef <- liftEffect $ Ref.new false
+    unless txInMempool do
+      -- Check if the transaction was included in the block:
+      txConfirmed <- checkTxByHashAff datumCacheWs lvl txHash
+      logger "Tx confirmed" txConfirmed txHash
+      unless txConfirmed $ liftEffect do
+        Ref.write true retrySubmitTxRef
+        sendRequest (requestBody /\ unit)
+    retrySubmitTx <- liftEffect $ Ref.read retrySubmitTxRef
```

Comment:
A suggestion for @firefrorefiddle's criticism 
```suggestion
    retrySubmitTx <-
      if txInMempool then do
        -- Check if the transaction was included in the block:
        txConfirmed <- checkTxByHashAff datumCacheWs lvl txHash
        logger "Tx confirmed" txConfirmed txHash
        unless txConfirmed $ liftEffect do
          sendRequest (requestBody /\ unit)
        pure (not txConfirmed)
      else pure false
```

Reply:
Actually it's probably better to use `MonadPlus`'s guard

Reply:
@jy14898 Thanks for this suggestion!